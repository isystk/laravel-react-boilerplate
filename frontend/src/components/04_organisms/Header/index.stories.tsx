import { Meta, Story } from '@storybook/react'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import React from 'react'
import Header, { HeaderProps } from './index'
import MainService from '@/services/main'

export default {
  title: '04_organisms/Header',
  component: Header,
} as Meta

const Template: Story = () => {
  const main = new MainService(() => ({}))
  const props: HeaderProps = {
    isMenuOpen: true,
    setMenuOpen: () => ({}),
    menuItems: [
      {
        label: 'Menu1',
        href: 'https://menu1',
      },
      {
        label: 'Menu2',
        href: 'https://menu2',
      },
      {
        label: 'Menu3',
        href: 'https://menu3',
        target: '_blank',
      },
    ],
  }
  if (!main) <></>
  return (
    <Context.Provider value={main}>
      <Header {...props} />
    </Context.Provider>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
