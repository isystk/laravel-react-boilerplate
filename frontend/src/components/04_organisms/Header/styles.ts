import { css, keyframes } from '@emotion/css'

export const header = css``
