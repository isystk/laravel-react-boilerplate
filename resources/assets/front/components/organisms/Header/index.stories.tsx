import Header from './index';
import { JSX, useEffect } from 'react';
import { AppProvider } from '@/states/AppContext';
import useAppRoot from '@/states/useAppRoot';
import { BrowserRouter } from 'react-router-dom';
import { Auth } from '@/states/auth';

export default {
  title: 'Components/Organisms/Header',
  component: Header,
  tags: ['autodocs'],
  decorators: [
    Story => (
      <BrowserRouter>
        <AppProvider>
          <Story />
        </AppProvider>
      </BrowserRouter>
    ),
  ],
};

export const Default: { render: () => null | JSX.Element } = {
  render: () => {
    const [state] = useAppRoot();
    if (!state) return null;

    return <Header />;
  },
};

export const Login: { render: () => null | JSX.Element } = {
  render: () => {
    const [state, service] = useAppRoot();

    useEffect(() => {
      service.auth.setAuth({
        id: 1,
        name: 'ユーザー名',
        email: 'test@test.com',
        email_verified_at: '2020-01-01 00:00:00',
      } as Auth);
    }, []);

    return <>{state && <Header />}</>;
  },
};
