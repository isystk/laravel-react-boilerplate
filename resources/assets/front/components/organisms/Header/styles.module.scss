@use "@/assets/styles/variable" as *;

.header {
    background-color: $white;
}

.menuContainer {
    margin-left: auto;
    ul {
        display: flex;
        justify-content: center;
        align-items: center;
        li {
            display: inline-block;
            margin-right: 12px;
        }
    }
}

.register-link {
    color: #e11d48;
    text-decoration: none;
}

.register-link:hover {
    text-decoration: underline;
}

.cart-link {
    margin-left: 12px;
}

.cart-image {
    height: 40px;
}

.contact-link {
    color: #e11d48;
    text-decoration: none;
}

.contact-link:hover {
    text-decoration: underline;
}

