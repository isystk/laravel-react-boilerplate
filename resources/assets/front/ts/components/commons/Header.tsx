import React, { FC, useState } from "react";
import {
    Dropdown,
    DropdownToggle,
    DropdownMenu,
    DropdownItem,
    Form,
} from "reactstrap";
import CSRFToken from "@/components/elements/CSRFToken";
import { Url } from "@/constants/url";
import Logo from "@/components/commons/Logo";
import { Link, useNavigate } from "react-router-dom";
import MainService from "@/services/main";

type Props = {
    appRoot: MainService;
};

export const CommonHeader: FC<Props> = ({ appRoot }) => {
    const navigate = useNavigate();
    const push_mycart = () => navigate(Url.MYCART);

    const [isDropdownOpen, setDropdownOpen] = useState(false);
    const [isSideMenu, setSideMenu] = useState(false);
    const { isLogined, name } = appRoot.auth;

    const renderLoginPc = (): JSX.Element => {
        return (
            <div className="" id="navbarSupportedContent">
                <ul className="navbar-nav ml-auto">
                    {(() => {
                        if (isLogined) {
                            // ログイン済みの場合
                            return (
                                <>
                                    <Dropdown
                                        isOpen={isDropdownOpen}
                                        toggle={() => {
                                            setDropdownOpen(!isDropdownOpen);
                                        }}
                                    >
                                        <DropdownToggle
                                            caret
                                            nav
                                            id="logout-nav"
                                        >
                                            {name + " 様"}
                                        </DropdownToggle>
                                        <DropdownMenu>
                                            <DropdownItem
                                                onClick={(e) => {
                                                    e.preventDefault();
                                                    const element: HTMLFormElement =
                                                        document.getElementById(
                                                            "logout-form"
                                                        ) as HTMLFormElement;
                                                    if (element) {
                                                        element.submit();
                                                    }
                                                }}
                                            >
                                                ログアウト
                                            </DropdownItem>
                                            <Form
                                                id="logout-form"
                                                action={Url.LOGOUT}
                                                method="POST"
                                                style={{ display: "none" }}
                                            >
                                                <CSRFToken appRoot={appRoot} />
                                            </Form>
                                            <DropdownItem
                                                onClick={mycartSubmit}
                                            >
                                                カートを見る
                                            </DropdownItem>
                                        </DropdownMenu>
                                        <Form
                                            id="mycart-form"
                                            action={Url.MYCART}
                                            method="POST"
                                            style={{ display: "none" }}
                                        >
                                            <CSRFToken appRoot={appRoot} />
                                        </Form>
                                    </Dropdown>

                                    <a href="#" onClick={mycartSubmit}>
                                        <img
                                            src="/assets/front/image/cart.png"
                                            className="cartImg ml-3"
                                        />
                                    </a>
                                </>
                            );
                        } else {
                            // 未ログインの場合
                            return (
                                <>
                                    <li className="nav-item">
                                        <Link
                                            className="btn btn-danger mr-3"
                                            to={Url.LOGIN}
                                        >
                                            ログイン
                                        </Link>
                                    </li>
                                    <li className="nav-item">
                                        <Link
                                            className="btn btn-link text-danger"
                                            to={Url.REGISTER}
                                        >
                                            新規登録
                                        </Link>
                                    </li>
                                </>
                            );
                        }
                    })()}

                    <li className="nav-item">
                        <Link
                            className="btn btn-link text-danger"
                            to={Url.CONTACT}
                        >
                            お問い合わせ
                        </Link>
                    </li>
                </ul>
            </div>
        );
    };

    const renderLoginSp = (): JSX.Element => {
        const isSideMenuOpen = isSideMenu;
        const sideMenuClass = isSideMenuOpen ? "open" : "";
        const menuBtnClass = isSideMenuOpen ? "menu-btn on" : "menu-btn";
        const layerPanelClass = isSideMenuOpen ? "on" : "";

        return (
            <>
                <div
                    className={menuBtnClass}
                    onClick={(e) => {
                        e.preventDefault();
                        const isOpen = isSideMenu ? false : true;
                        setSideMenu(isOpen);
                    }}
                >
                    <figure></figure>
                    <figure></figure>
                    <figure></figure>
                </div>
                <div id="side-menu" className={sideMenuClass}>
                    <div className="side-menu-header">
                        <p style={{ margin: "20px", fontSize: "16px" }}>
                            {isLogined && name + " 様"}
                        </p>
                    </div>
                    <nav>
                        <ul>
                            {(() => {
                                if (isLogined) {
                                    // ログイン済みの場合
                                    return (
                                        <>
                                            <li>
                                                <a
                                                    href="#"
                                                    onClick={(e) => {
                                                        e.preventDefault();
                                                        const element: HTMLFormElement =
                                                            document.getElementById(
                                                                "logout-form"
                                                            ) as HTMLFormElement;
                                                        if (element) {
                                                            element.submit();
                                                        }
                                                    }}
                                                >
                                                    ログアウト
                                                </a>
                                                <Form
                                                    id="logout-form"
                                                    action={Url.LOGOUT}
                                                    method="POST"
                                                    style={{ display: "none" }}
                                                >
                                                    <CSRFToken
                                                        appRoot={appRoot}
                                                    />
                                                </Form>
                                            </li>
                                            <li>
                                                <a
                                                    href="#"
                                                    onClick={mycartSubmit}
                                                >
                                                    カートを見る
                                                </a>
                                                <Form
                                                    id="mycart-form"
                                                    action={Url.MYCART}
                                                    method="POST"
                                                    style={{ display: "none" }}
                                                >
                                                    <CSRFToken
                                                        appRoot={appRoot}
                                                    />
                                                </Form>
                                            </li>
                                        </>
                                    );
                                } else {
                                    // 未ログインの場合
                                    return (
                                        <>
                                            <li>
                                                <Link
                                                    to={Url.LOGIN}
                                                    onClick={() => {
                                                        setSideMenu(false);
                                                    }}
                                                >
                                                    ログイン
                                                </Link>
                                            </li>
                                            <li>
                                                <Link
                                                    to={Url.REGISTER}
                                                    onClick={() => {
                                                        setSideMenu(false);
                                                    }}
                                                >
                                                    新規登録
                                                </Link>
                                            </li>
                                        </>
                                    );
                                }
                            })()}
                            <li>
                                <Link
                                    to={Url.CONTACT}
                                    onClick={() => {
                                        setSideMenu(false);
                                    }}
                                >
                                    お問い合わせ
                                </Link>
                            </li>
                        </ul>
                    </nav>
                </div>
                <div id="layer-panel" className={layerPanelClass}></div>
            </>
        );
    };

    const mycartSubmit = (e) => {
        e.preventDefault();
        setSideMenu(false);
        push_mycart();
    };

    return (
        <>
            <header className="header shadow-sm">
                <nav className="navbar navbar-expand-md navbar-light bg-white headerNav">
                    <Logo />
                    {renderLoginPc()}
                    {renderLoginSp()}
                </nav>
            </header>
        </>
    );
};

export default CommonHeader;
