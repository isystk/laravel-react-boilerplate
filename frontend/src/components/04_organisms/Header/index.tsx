import React, { FC, useEffect, useState } from 'react'
import { ContainerProps } from 'types'
import { connect } from '@/components/hoc'
import Logo from '@/components/01_atoms/Logo'
import Hamburger from '@/components/01_atoms/Hamburger'
import * as styles from './styles'
import { useI18n } from '@/components/i18n'
import { MenuItem } from '@/constants/menu'

/** HeaderProps Props */
export type HeaderProps = {
  isMenuOpen: boolean
  setMenuOpen: () => void
  menuItems: MenuItem[]
  scrollY: number
}
/** Presenter Props */
export type PresenterProps = HeaderProps & {
  t
}

/** Presenter Component */
const HeaderPresenter: FC<PresenterProps> = ({
  isMenuOpen,
  setMenuOpen,
  menuItems = [],
  t,
  isVisible,
}) => (
  <>
    <header
      className={`${styles.header} ${
        isVisible ? 'fixed' : 'bg-main'
      } z-50 flex justify-between items-center h-16 px-4 py-4 sm:px-8 w-full`}
    >
      <Logo />
      <div
        className={`flex pr-3 z-50 md:hidden ${
          isMenuOpen ? 'fixed right-5' : ''
        }`}
      >
        <Hamburger isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
      </div>
      <nav className="hidden sm:block">
        <ul className="flex gap-6">
          {menuItems.map(({ label, href, target }, idx) => (
            <li key={idx}>
              <a href={href} target={target} rel={target ? 'noreferrer' : ''}>
                {t(label)}
              </a>
            </li>
          ))}
        </ul>
      </nav>
    </header>
  </>
)

/** Container Component */
const HeaderContainer: React.FC<
  ContainerProps<HeaderProps, PresenterProps>
> = ({ presenter, scrollY = 500, ...props }) => {
  const { t } = useI18n('Common')
  const [isVisible, setIsVisible] = useState(false)

  useEffect(() => {
    const toggleVisibility = () => {
      if (window.scrollY > scrollY) {
        setIsVisible(true)
      } else {
        setIsVisible(false)
      }
    }

    window.addEventListener('scroll', toggleVisibility)

    return () => window.removeEventListener('scroll', toggleVisibility)
  }, [scrollY])

  return presenter({
    t,
    isVisible,
    ...props,
  })
}

export default connect<HeaderProps, PresenterProps>(
  'Header',
  HeaderPresenter,
  HeaderContainer
)
