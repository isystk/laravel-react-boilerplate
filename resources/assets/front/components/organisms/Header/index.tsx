import styles from './styles.module.scss';
import Logo from '@/components/atoms/Logo';
import { Link, useNavigate } from 'react-router-dom';
import { Url } from '@/constants/url';
import Image from '@/components/atoms/Image';
import DropDown from '@/components/atoms/DropDown';
import SideMenu from '@/components/organisms/SideMenu';
import useAppRoot from '@/states/useAppRoot';
import CSRFToken from '@/components/atoms/CSRFToken';
import cartImage from '@/assets/images/cart.png';

const Header = () => {
  const { state } = useAppRoot();
  if (!state) return <></>;

  const { isLogined, name } = state.auth;

  const navigate = useNavigate();
  return (
    <header className={`${styles.header} shadow-sm`}>
      <nav className="flex flex-wrap items-center justify-between bg-white px-4 py-3">
        <Logo />
        {/* メニュー（PC表示） */}
        <div className={`${styles.menuContainer} hidden md:flex`}>
          <ul>
            {(() => {
              if (isLogined) {
                return (
                  <li>
                    <DropDown
                      text={`${name} 様`}
                      items={[
                        {
                          text: 'ログアウト',
                          onClick: () => {
                            const element: HTMLFormElement = document.getElementById(
                              'logout-form',
                            ) as HTMLFormElement;
                            if (element) {
                              element.submit();
                            }
                          },
                        },
                        {
                          text: 'カートを見る',
                          onClick: () => navigate(Url.myCart),
                        },
                      ]}
                    />
                  </li>
                );
              } else {
                return (
                  <>
                    <li>
                      <Link className="btn btn-danger" to={Url.login}>
                        ログイン
                      </Link>
                    </li>
                    <li>
                      <Link to={Url.register}>新規登録</Link>
                    </li>
                  </>
                );
              }
            })()}
            <li>
              <Link to={Url.myCart}>
                <Image src={cartImage as string} width={30} height={30} alt="カート" />
              </Link>
            </li>
            <li>
              <Link to={Url.contact}>お問い合わせ</Link>
            </li>
          </ul>
        </div>
        {/* サイドメニュー */}
        <SideMenu
          text={isLogined ? `${name} 様` : ''}
          items={(() => {
            const items: Array<{ text: string; onClick?: () => void }> = [];
            if (isLogined) {
              items.push({
                text: 'ログアウト',
                onClick: () => {
                  const element: HTMLFormElement = document.getElementById(
                    'logout-form',
                  ) as HTMLFormElement;
                  if (element) {
                    element.submit();
                  }
                },
              });
            } else {
              items.push(
                {
                  text: 'ログイン',
                  onClick: () => navigate(Url.login),
                },
                {
                  text: '新規登録',
                  onClick: () => navigate(Url.register),
                },
              );
            }
            items.push(
              {
                text: 'カートを見る',
                onClick: () => navigate(Url.myCart),
              },
              {
                text: 'お問い合わせ',
                onClick: () => navigate(Url.contact),
              },
            );
            return items;
          })()}
        />
      </nav>
      <form id="logout-form" action={Url.logout} method="POST">
        <CSRFToken />
      </form>
    </header>
  );
};

export default Header;
