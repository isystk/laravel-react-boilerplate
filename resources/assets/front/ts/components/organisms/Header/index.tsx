import styles from './styles.module.scss';
import Logo from "@/components/molecules/Logo";
import {Link, useNavigate} from "react-router-dom";
import { Url } from "@/constants/url";
import Image from "@/components/atoms/Image";
import DropDown from "@/components/atoms/DropDown";
import SideMenu from "@/components/atoms/SideMenu";

const Header = () => {
    const navigate = useNavigate();
    return (
        <header className={`${styles.header} shadow-sm`}>
            <nav className="flex flex-wrap items-center justify-between bg-white px-4 py-3">
                {/* ロゴ */}
                <Logo />
                {/* メニュー（PC表示） */}
                <div className={`${styles.menuContainer} hidden md:flex`}>
                    <ul>
                        <li>
                            <Link className="btn btn-danger" to={Url.LOGIN}>ログイン</Link>
                        </li>
                        <li>
                            <DropDown
                                text={"テスト 様"}
                                items={[
                                    {
                                        text: "ログアウト",
                                        onClick: () => {
                                            const element: HTMLFormElement = document.getElementById("logout-form") as HTMLFormElement;
                                            if (element) {
                                                element.submit();
                                            }
                                        }
                                    },
                                    {
                                        text: "カートを見る",
                                        onClick: () => navigate(Url.MYCART)
                                    }
                                ]}
                            />
                        </li>
                        <li>
                            <Link to={Url.REGISTER}>新規登録</Link>
                        </li>
                        <li>
                            <Link
                                to={Url.MYCART}
                            >
                                <Image src="https://localhost/assets/front/image/cart.png" width={30} height={30} alt="カート"/>
                            </Link>
                        </li>
                        <li>
                            <Link to={Url.CONTACT}>お問い合わせ</Link>
                        </li>
                    </ul>
                </div>

                {/* サイドメニュー */}
                <SideMenu
                    items={[
                        {
                            text: "ログイン",
                            onClick: () => navigate(Url.LOGIN)
                        },
                        {
                            text: "新規登録",
                            onClick: () => navigate(Url.REGISTER)
                        },
                        {
                            text: "ログアウト",
                            onClick: () => {
                                const element: HTMLFormElement = document.getElementById("logout-form") as HTMLFormElement;
                                if (element) {
                                    element.submit();
                                }
                            }
                        },
                        {
                            text: "カートを見る",
                            onClick: () => navigate(Url.MYCART)
                        },
                        {
                            text: "お問い合わせ",
                            onClick: () => navigate(Url.CONTACT)
                        }
                    ]}
                />
            </nav>
        </header>
    );
};

export default Header;
