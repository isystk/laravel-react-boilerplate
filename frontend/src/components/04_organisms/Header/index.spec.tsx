import React from 'react'
import { render } from '@testing-library/react'
import Header, { HeaderProps } from './index'

describe('LandingPageTemplate', () => {
  it('Match Snapshot', () => {
    const props: HeaderProps = {
      isMenuOpen: true,
      setMenuOpen: () => ({}),
      menuItems: [
        {
          label: 'Menu1',
          href: 'https://menu1',
        },
        {
          label: 'Menu2',
          href: 'https://menu2',
        },
        {
          label: 'Menu3',
          href: 'https://menu3',
          target: '_blank',
        },
      ],
    }
    const { asFragment } = render(<Header {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
