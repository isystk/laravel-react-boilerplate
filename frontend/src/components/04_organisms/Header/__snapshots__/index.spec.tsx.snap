// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingPageTemplate Match Snapshot 1`] = `
<DocumentFragment>
  <header
    class="css-0 bg-main z-50 flex justify-between items-center h-16 px-4 py-4 sm:px-8 w-full"
  >
    <a
      class="css-0"
      href="#"
    >
      <div
        class="w-20 md:w-40"
      >
        <img
          class="css-s08p0c "
          src="/images/logo.png"
        />
      </div>
    </a>
    <div
      class="flex pr-3 z-50 md:hidden fixed right-5"
    >
      <button
        class="space-y-2 css-jgxxba"
      >
        <div
          class="w-6 h-0.5 bg-black translate-y-2.5 rotate-45"
        />
        <div
          class="w-6 h-0.5 bg-black -rotate-45"
        />
        <div
          class="opacity-0"
        />
      </button>
    </div>
    <nav
      class="hidden sm:block"
    >
      <ul
        class="flex gap-6"
      >
        <li>
          <a
            href="https://menu1"
            rel=""
          >
            Menu1
          </a>
        </li>
        <li>
          <a
            href="https://menu2"
            rel=""
          >
            Menu2
          </a>
        </li>
        <li>
          <a
            href="https://menu3"
            rel="noreferrer"
            target="_blank"
          >
            Menu3
          </a>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;
